lib/sax.js

```diff
@@ -651,16 +649,22 @@
 	//	getType:function(i){},
 };
 
-function split(source, start) {
-	var match;
-	var buf = [];
-	var reg = /'[^']+'|"[^"]+"|[^\s<>\/=]+=?|(\/?\s*>|<)/g;
-	reg.lastIndex = start;
-	reg.exec(source); //skip <
-	while ((match = reg.exec(source))) {
-		buf.push(match);
-		if (match[1]) return buf;
+function _experimentalDoctypeRegex(source, start) {
+	// since 0.5.0 we purposefully do not remove the quotes from publicId and systemId
+	// so the values can be serialized as is
+	var regex =
+		/<!DOCTYPE\s+(([:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD]|[\uD800-\uDB7F\uDC00-\uDFFF])([:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\-\.0-9\u00B7\u0300-\u036F\u203F-\u2040]|[\uD800-\uDB7F\uDC00-\uDFFF])*)(\s+(SYSTEM)\s+("[^"]*"|'[^']*')|\s+(PUBLIC)\s+("[\r\n\x20\w\-'()+,.\/:=?;!*#@$_%]*"|'[\r\n\x20\w\-()+,.\/:=?;!*#@$_%]*')\s+("[^"]*"|'[^']*'))?\s*(\[((\s*(%(([:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD]|[\uD800-\uDB7F\uDC00-\uDFFF])([:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\-\.0-9\u00B7\u0300-\u036F\u203F-\u2040]|[\uD800-\uDB7F\uDC00-\uDFFF])*);|<!ELEMENT ([^>]*)>|<!ATTLIST ([^>]*)>|<!ENTITY ([^>]*)>|<!NOTATION ([^>]*)>|<\?([^>]*)\?>|<!--([^>]*)-->)\s*)*)])?\s*>/;
```

Comment:
@shunkica we already have the `NameStart` and `NameChar` in the `conventions` module with an example of composing a regular expression from the source of another one.
Would be nice to use the same approach here to make it more readable.

I'm not sure if that is what the CodeScanning complains about though.

Reply:
I will look into that.

Reply:
I know how impressive it is to have such a mighty regular expression, but the amount of time I already invested trying to understand it, tells me it's to complex to maintain in a reasonable way.
I would suggest to split it into multiple pieces that are wrapped by some conditionals, to make it possible to understand what is happening.
e.g. on for getting publicId/systemId, another one for the rest. or if possible split it even further.

Reply:
> I know how impressive it is to have such a mighty regular expression, but the amount of time I already invested trying to understand it, tells me it's to complex to maintain in a reasonable way. I would suggest to split it into multiple pieces that are wrapped by some conditionals, to make it possible to understand what is happening. e.g. on for getting publicId/systemId, another one for the rest. or if possible split it even further.

My first approach was tokenizing it but that seemed even worse. It was just lines and lines of if then conditions and after a while I got lost.

I understand that this sausage looks bad, but if it is up to spec I don't see it as something that should necessarily be maintained. Also, tools like [this](https://regex101.com/r/oDPk1V/2) which help visualize the regex make it not look as bad. Its mostly just all the Unicode sequences that are ugly.

If you feel strongly about this I will try something else.

Reply:
@karfau Would you consider adding [xregexp](https://github.com/slevithan/xregexp) as a dependancy?
It allows for atomic groups which would solve the backtracking issue reported by CodeQL.
It also allows some other stuff like named capture groups which make the regexp more readable.

Reply:
So far we do support ES5, which doesn't come with named capture groups.
And we do not use any dependency at runtime, so the step of introducing one is a big one.
The library looks very impressive.
From the docs it sounds like it might be possible to use it as a dev dependency to create the regular expression as a build step, but I'm not really sure about that.

And yes I feel strongly about this, because currently I'm the only active/core maintainer. So at least I have to be able to understand those regular expressions. Otherwise should somebody find any performance or security issue with it, I'm lost.

Whether that means it needs to be split or we need to introduce a dependency or a build step, or what ever it takes, that is something we have to figure out.
Are you up to having a call to discuss it in a sync way instead of going back and forth with comments?
If yes, you can find my mail on my profile to suggest a tool for connecting of your choice.